File: 2017/elixir/lib/day11/hexgrid.ex
require IEx

defmodule Advent2017.Day11 do
  @doc ~S"""
      iex> Advent2017.Day11.track("ne,ne,ne")
      ...> |> List.first
      ...> |> Advent2017.Day11.distance
      3
      iex> Advent2017.Day11.track("ne,ne,sw,sw")
      ...> |> List.first
      ...> |> Advent2017.Day11.distance
      0
      iex> Advent2017.Day11.track("ne,ne,s,s")
      ...> |> List.first
      ...> |> Advent2017.Day11.distance
      2
      iex> Advent2017.Day11.track("se,sw,se,sw,sw")
      ...> |> List.first
      ...> |> Advent2017.Day11.distance
      3
  """
  def track(steps) do
    steps
    |> String.trim
    |> String.split(",", [trim: true])
    |> Enum.map(&String.to_atom &1)
    |> track([%{x: 0, y: 0, z: 0}])
  end

  def track([], states), do: states
  def track([dir|future], states) do
    track(future, [step(dir, hd(states))|states])
  end

  def step(:n,  c), do: %{c | y: c[:y]+1, z: c[:z]-1}
  def step(:ne, c), do: %{c | x: c[:x]+1, z: c[:z]-1}
  def step(:se, c), do: %{c | x: c[:x]+1, y: c[:y]-1}
  def step(:s,  c), do: %{c | z: c[:z]+1, y: c[:y]-1}
  def step(:sw, c), do: %{c | z: c[:z]+1, x: c[:x]-1}
  def step(:nw, c), do: %{c | y: c[:y]+1, x: c[:x]-1}

  def distance(state) do
    Map.values(state)
    |> Enum.map(&abs &1)
    |> Enum.sum
    |> div(2)
  end

  def p1 do
    {:ok, file} = File.read(__DIR__ <> "/input.txt")
    file
    |> track
    |> List.first
    |> distance
  end

  def p2 do
    {:ok, file} = File.read(__DIR__ <> "/input.txt")
    file
    |> track
    |> Enum.map(&distance &1)
    |> Enum.max
  end
end

File: 2017/ruby/day11/hexgrid.rb
require 'minitest/autorun'

class Position
  attr_accessor :x, :y, :z, :history

  def initialize
    @x = 0
    @y = 0
    @z = 0

    @history = []
  end

  def inspect
    "X: #{@x}, Y: #{@y}, Z: #{@z}\n"
  end

  def n
    @history.push self.clone

    @y += 1
    @z -= 1
  end

  def ne
    @history.push self.clone

    @x += 1
    @z -= 1
  end

  def se
    @history.push self.clone

    @x += 1
    @y -= 1
  end

  def s
    @history.push self.clone

    @z += 1
    @y -= 1
  end

  def sw
    @history.push self.clone

    @z += 1
    @x -= 1
  end

  def nw
    @history.push self.clone

    @y += 1
    @x -= 1
  end

  def distance
    (@x.abs + @y.abs + @z.abs)/2
  end
end

def track(instructions)
  position = Position.new

  instructions.each do |direction|
    position.send(direction)
  end

  position
end

def p1
  instructions = File.read("./input.txt").strip.split(",")
  track(instructions).distance
end

def p2
  instructions = File.read("./input.txt").strip.split(",")
  track(instructions).history.map{|x| x.distance }.max
end

puts "Part 1:"
puts p1
puts "Part 2:"
puts p2

class HexTest < Minitest::Test
  def test_distance_1
    assert_equal 3, track(["ne","ne","ne"]).distance
  end
  def test_distance_2
    assert_equal 0, track(["ne","ne","sw","sw"]).distance
  end
  def test_distance_3
    assert_equal 2, track(["ne","ne","s","s"]).distance
  end
  def test_distance_4
    assert_equal 3, track(["se","sw","se","sw","sw"]).distance
  end
end
